pkg/assetfetcher/http_test.go

```diff
@@ -0,0 +1,642 @@
+			if !ok {
+				errChan = nil
+			} else {
+				actualError = err
+			}
+		}
+		if assetChan == nil && errChan == nil {
+			break
+		}
+	}
+	assert.IsType(t, &URLParsingError{}, actualError)
+}
+
+func TestMakeRequestSuccess(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockRT := NewMockRoundTripper(ctrl)
+
+	asset := domain.Asset{
+		IP: "127.0.0.1",
+		ID: 123456,
+	}
+	resp := SiteAssetsResponse{
+		Resources: []domain.Asset{asset},
+		Page:      Page{},
+		Links:     domain.Link{},
+	}
+
+	respJSON, _ := json.Marshal(resp)
+	mockRT.EXPECT().RoundTrip(gomock.Any()).Return(&http.Response{
+		Body:       ioutil.NopCloser(bytes.NewReader(respJSON)),
+		StatusCode: http.StatusOK,
+	}, nil)
+
+	assetFetcher := &NexposeAssetFetcher{
+		HTTPClient: &http.Client{Transport: mockRT},
+		Host:       "http://localhost",
+		PageSize:   100,
+	}
+
+	var wg sync.WaitGroup
+	assetChan := make(chan domain.Asset, 1)
+	errChan := make(chan error, 1)
+	defer close(assetChan)
+	defer close(errChan)
+
+	wg.Add(1)
+	assetFetcher.makeRequest(context.Background(), &wg, "siteID", 100, assetChan, errChan)
+
+	assert.NotNil(t, <-assetChan)
+}
+
+func TestMakeRequestWithBadResponse(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Length", "1")
+	}))
+	defer ts.Close()
+	assetFetcher := &NexposeAssetFetcher{
+		HTTPClient: ts.Client(),
+		Host:       ts.URL,
+		PageSize:   100,
+	}
+
+	var wg sync.WaitGroup
+	assetChan := make(chan domain.Asset, 1)
+	errChan := make(chan error, 1)
+
+	wg.Add(1)
+	assetFetcher.makeRequest(context.Background(), &wg, "siteID", 100, assetChan, errChan)
+	wg.Wait()
+
+	close(assetChan)
+	close(errChan)
+
+	assert.IsType(t, &ErrorReadingNexposeResponse{}, <-errChan)
+}
+
+func TestMakeRequestWithInvalidHost(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockRT := NewMockRoundTripper(ctrl)
+	resp := SiteAssetsResponse{
+		Resources: []domain.Asset{},
+		Page:      Page{},
+		Links:     domain.Link{},
+	}
+	respJSON, _ := json.Marshal(resp)
+	mockRT.EXPECT().RoundTrip(gomock.Any()).Return(&http.Response{
+		Body:       ioutil.NopCloser(bytes.NewReader(respJSON)),
+		StatusCode: http.StatusOK,
+	}, nil).Times(0)
+
+	assetFetcher := &NexposeAssetFetcher{
+		HTTPClient: &http.Client{Transport: mockRT},
+		Host:       "~!@#$%^&*()_+:?><!@#$%^&*())_:",
+		PageSize:   100,
+	}
+
+	var wg sync.WaitGroup
+	assetChan := make(chan domain.Asset, 1)
+	errChan := make(chan error, 1)
+
+	wg.Add(1)
+	assetFetcher.makeRequest(context.Background(), &wg, "siteID", 100, assetChan, errChan)
+	wg.Wait()
+
+	close(assetChan)
+	close(errChan)
+
+	assert.IsType(t, &URLParsingError{}, <-errChan)
+}
+
+func TestMakeRequestHTTPError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockRT := NewMockRoundTripper(ctrl)
+
+	mockRT.EXPECT().RoundTrip(gomock.Any()).Return(nil, errors.New("HTTPError"))
```

Comment:
Given the conversation about testing `ioutil.ReadAll` related failure cases from the previous version of this code, I think I am confused as to the intent of this test. This test covers what I would expect it to given the name `HTTPError` because it will trigger returning an error from the `http.Client.Do` method. However, the question you asked previously was how to test the code path related to getting a successful HTTP response that later suffers an error while reading from the response body.

Personally, I'm OK with not having the `ReadAll` path covered but I wanted to check that you intended not to test it first.

Reply:
You might be thinking of 2 different tests. The test that tests for when an error is returned from `ioutil.ReadAll()` should result in a `ErrorReadingNexposeResponse` error, which is different than _this_ test, which is testing what happens when an error is returned from `http.Client.Do`.

I'll rename `TestMakeRequestWithBadResponse` to `TestMakeRequestWithErrorReadingResponse` for clarification